<div>
<h2 style="background-color:Purple;text-align:center;border-radius:40px;border: 1px;padding: 20px; color:white"><b>Filoger Hands On Prerequisite Python For AI <br>Exercises 07 & 08 | Deadline: 2023 9 May <br> Score: 700</h2>
</div>

# Name

### Meisam Kargar

<h3 style="background-color:#ff3300
;text-align:center;border-radius:40px;border: 1px;padding: 10px;color:white"><b>Exercises 09<br>Score: 550</h3>

<h5 style="background-color:#1765e3;text-align:left;border-radius:40px;border: 1px;padding: 10px;color:white"><b>1- Based on some examples (code), show the difference between list and tuple(50)</h5>

In [ ]:
# list:
list1 = [1,2,3,'hello',True]
list1[4] = 'goodby' # list function is changeable
print(list1)

# -----------------------------

# tuple:
touple1 = (1,2,3,'hello', True)
touple1[4] = 'goodby' # becuase touple is not changeable we have an error


<h5 style="background-color:#1765e3;text-align:left;border-radius:40px;border: 1px;padding: 10px;color:white"><b>2- Tuple(100)</h5>

Create an empty tuple

In [ ]:
t = ()  # This is an empty touple


Create a tuple with one item

In [ ]:
t1 = ('One item', )  # This is a one item touple


Create a tuple with more than one item.

In [ ]:
t2 = ('hello', 1, 2, True) # This is a multi items touple


Concatenate two tuples of Part2 and Part3

In [ ]:
t3 = t1 + t2 # concatate t1 with t2
print(f'concated t1 and t2: {t3}', end='\n')


Check 1 is in tuple of Part4 or no.

In [ ]:
if 1 in t3:
    print('yes is there')
else:
    print("no this value not in touple")

<h5 style="background-color:#1765e3;text-align:left;border-radius:40px;border: 1px;padding: 10px;color:white"><b>3- Answer the questions according to the list below.
numbers = (10, 20, 30, 5, 40, 20, 4, 2, 1,8, 80)  (100)</h5>

Add 100 at the end of tuple

In [1]:
numbers = (10, 20, 30, 5, 40, 20, 4, 2, 1, 8, 80)
numbers = numbers + (100, )

Add 50 in the index of 3.

In [2]:
numbers = numbers[0:3] + (50,) + numbers[3:]

Remove 30 from tuple

In [3]:
numbers = numbers[0:2] + numbers[3:]

Edit 10 with 110.

In [4]:
numbers =  (110, ) + numbers[1:]

Sort the tuple.

In [5]:
numbers = sorted(numbers)
numbers

[1, 2, 4, 5, 8, 20, 20, 40, 50, 80, 100, 110]

<h5 style="background-color:#1765e3;text-align:left;border-radius:40px;border: 1px;padding: 10px;color:white"><b>4- Write a python program that takes a list of tuple and display other list that only have
first and last items of each tuple.</h5>

In [6]:
# Input:[(1, 2, 3), (3, 5, 6), (4, 10, 12)]
# Output: [(1, 3), (3, 6), (4, 12)]

counter = 0
list = []
while counter != 3:
    inp = tuple(input('enter a number').split())
    list.append(inp)
    counter += 1
mid_remove = [i[0::2] for i in list]
print(mid_remove)

[('1', '3'), ('3', '6'), ('4', '12')]


<h5 style="background-color:#1765e3;text-align:left;border-radius:40px;border: 1px;padding: 10px;color:white"><b>5- Generate 10 different random numbers between 100 and two 200.(100)
 </h5>

In [7]:
import random
for i in range(10):
    random_numbers = random.sample(range(100, 201), 10)
    print(set(random_numbers))

{166, 137, 123, 108, 111, 180, 149, 153, 187, 127}
{132, 197, 198, 101, 168, 173, 153, 127, 158, 191}
{163, 164, 167, 105, 138, 139, 145, 114, 125, 127}
{194, 135, 104, 110, 176, 181, 150, 151, 152, 159}
{161, 101, 167, 104, 107, 111, 147, 148, 150, 122}
{128, 129, 195, 163, 164, 105, 109, 184, 122, 155}
{128, 131, 132, 101, 103, 200, 170, 150, 122, 127}
{161, 109, 174, 146, 182, 122, 187, 127, 189, 191}
{192, 195, 142, 178, 115, 149, 121, 123, 188, 158}
{130, 195, 163, 133, 199, 103, 150, 152, 158, 127}


<h5 style="background-color:#1765e3;text-align:left;border-radius:40px;border: 1px;padding: 10px;color:white"><b>6- Create two lists, then create a third list that contains values that are in the first list but
not in the second list.
 </h5>

In [8]:
l1 = [1,5,9,5,3]
l2 = [5,6,9,7,1]
l3 = []
l3 = l1 + l2
print(set(l3))

{1, 3, 5, 6, 7, 9}


<h5 style="background-color:#1765e3;text-align:left;border-radius:40px;border: 1px;padding: 10px;color:white"><b>7- Find two new methods for Tuple and Set.
 </h5>

In [11]:
dic1 = {1,2,5,6,30}
dic2 = {3,5,1,6,35}
# Union method
res1 = dic1 | dic2
print(res1)

# subscription method:
res2 = dic1 & dic2
print(res2)

{1, 2, 3, 35, 5, 6, 30}
{1, 5, 6}


In [13]:
Tuple = (0, 1, 2, 3, 2, 3, 1, 3, 2)
# getting the index of 3
res = Tuple.index(3)
print('First occurrence of 3 is', res)
 
Tuple_1 = (0, 1, (2, 3), (2, 3), 1,[3, 2], 'geeks', (0,))
# count the appearance of (2, 3)
res = Tuple_1.count((2, 3))
print('Count of (2, 3) in Tuple is:', res)


First occurrence of 3 is 3
Count of (2, 3) in Tuple is: 2


<h3 style="background-color:#ff3300
;text-align:center;border-radius:40px;border: 1px;padding: 10px;color:white"><b>Exercise 10<br>Score: 150</h3>

<h5 style="background-color:#1765e3;text-align:left;border-radius:40px;border: 1px;padding: 10px;color:white"><b>1- Write a recursive function for the binary search algorithm. (150)
 </h5>

In [9]:
def binary_search(array, low, high, x):
 
    
    if high >= low:
        mid = (high + low) // 2
 
        
        if array[mid] == x:
            return mid
 
        elif array[mid] > x:
            return binary_search(array, low, mid - 1, x)
 
        else:
            return binary_search(array, mid + 1, high, x)
 
    else:
        return -1

array = [ 30, 4, 6, 8, 20, 40, 60, 80]
x = 20

result = binary_search(array, 0, len(array)-1, x)
 
if result != -1:
    print("The index of the Element is", str(result))
else:
    print("This element is not present in your Array.")


The index of the Element is 4
